ispcrt/ispcrt.cpp

```diff
@@ -78,6 +87,321 @@ static OBJECT_T &referenceFromHandle(HANDLE_T handle) {
         return a;                                                                                                      \
     }
 
+
+// Define names of devices libraries.
+#if defined(_WIN32) || defined(_WIN64)
+#define ISPCRT_SO_LIB_PREFIX ""
+#define ISPCRT_SO_LIB_SUFFIX "dll"
+#elif defined(__APPLE__)
+#define ISPCRT_SO_LIB_PREFIX "lib"
+#define ISPCRT_SO_LIB_SUFFIX "dylib"
+#else
+#define ISPCRT_SO_LIB_PREFIX "lib"
+#define ISPCRT_SO_LIB_SUFFIX "so"
+#endif
+
+#define ISPCRT_DEVICE_CPU_SOLIB_PREFIX ISPCRT_SO_LIB_PREFIX "ispcrt_device_cpu."
+#define ISPCRT_DEVICE_GPU_SOLIB_PREFIX ISPCRT_SO_LIB_PREFIX "ispcrt_device_gpu."
+#define ISPCRT_DEVICE_CPU_SOLIB_NAME ISPCRT_DEVICE_CPU_SOLIB_PREFIX ISPCRT_SO_LIB_SUFFIX
+#define ISPCRT_DEVICE_GPU_SOLIB_NAME ISPCRT_DEVICE_GPU_SOLIB_PREFIX ISPCRT_SO_LIB_SUFFIX
+
+#if defined(_WIN32) || defined(_WIN64)
+#define ISPCRT_DEVICE_CPU_SOLIB_MAJOR_VERSION_NAME nullptr
+#define ISPCRT_DEVICE_GPU_SOLIB_MAJOR_VERSION_NAME nullptr
+#define ISPCRT_DEVICE_CPU_SOLIB_FULL_VERSION_NAME  nullptr
+#define ISPCRT_DEVICE_GPU_SOLIB_FULL_VERSION_NAME  nullptr
+#elif defined(__APPLE__)
+#define ISPCRT_DEVICE_CPU_SOLIB_MAJOR_VERSION_NAME \
+    ISPCRT_DEVICE_CPU_SOLIB_PREFIX "." ISPCRT_VERSION_MAJOR "." ISPCRT_SO_LIB_SUFFIX
+#define ISPCRT_DEVICE_GPU_SOLIB_MAJOR_VERSION_NAME \
+    ISPCRT_DEVICE_GPU_SOLIB_PREFIX "." ISPCRT_VERSION_MAJOR "." ISPCRT_SO_LIB_SUFFIX
+#define ISPCRT_DEVICE_CPU_SOLIB_FULL_VERSION_NAME \
+    ISPCRT_DEVICE_CPU_SOLIB_PREFIX "." ISPCRT_VERSION_FULL "." ISPCRT_SO_LIB_SUFFIX
+#define ISPCRT_DEVICE_GPU_SOLIB_FULL_VERSION_NAME \
+    ISPCRT_DEVICE_GPU_SOLIB_PREFIX "." ISPCRT_VERSION_FULL "." ISPCRT_SO_LIB_SUFFIX
+#else
+#define ISPCRT_DEVICE_CPU_SOLIB_MAJOR_VERSION_NAME ISPCRT_DEVICE_CPU_SOLIB_NAME "." ISPCRT_VERSION_MAJOR
+#define ISPCRT_DEVICE_GPU_SOLIB_MAJOR_VERSION_NAME ISPCRT_DEVICE_GPU_SOLIB_NAME "." ISPCRT_VERSION_MAJOR
+#define ISPCRT_DEVICE_CPU_SOLIB_FULL_VERSION_NAME ISPCRT_DEVICE_CPU_SOLIB_NAME "." ISPCRT_VERSION_FULL
+#define ISPCRT_DEVICE_GPU_SOLIB_FULL_VERSION_NAME ISPCRT_DEVICE_GPU_SOLIB_NAME "." ISPCRT_VERSION_FULL
+#endif
+
+// OS agnostic function to dynamically load a shared library.
+void *dyn_load_lib(const char *name, const char *name_major_version, const char *name_full_version) {
+#if defined(_WIN32) || defined(_WIN64)
+    return LoadLibrary(name);
+#else
+    void *handle = dlopen(name_full_version, RTLD_NOW | RTLD_LOCAL);
```

Comment:
It's clear what you're doing here, but still would be good to add some comments explaining why do you search for libraries in this order.